src/engine/BoundsExplainer.cpp

```diff
@@ -0,0 +1,270 @@
+		{
+			_lowerBoundsExplanations[i].clear();
+			_lowerBoundsExplanations[i].resize( other._lowerBoundsExplanations[i].size() );
+			_lowerBoundsExplanations[i].shrink_to_fit();
+		}
+
+		std::copy( other._upperBoundsExplanations[i].begin(), other._upperBoundsExplanations[i].end(), _upperBoundsExplanations[i].begin() );
+		std::copy( other._lowerBoundsExplanations[i].begin(), other._lowerBoundsExplanations[i].end(), _lowerBoundsExplanations[i].begin() );
+	}
+
+	return *this;
+}
+
+const std::vector<double>& BoundsExplainer::getExplanation( const unsigned var, const bool isUpper )
+{
+	ASSERT ( var < _varsNum );
+	return isUpper ? _upperBoundsExplanations[var] : _lowerBoundsExplanations[var];
+}
+
+void BoundsExplainer::updateBoundExplanation( const TableauRow& row, const bool isUpper )
+{
+	if ( !row._size )
+		return;
+	updateBoundExplanation( row, isUpper, row._lhs );
+}
+
+void BoundsExplainer::updateBoundExplanation( const TableauRow& row, const bool isUpper, const unsigned var )
+{
+	if ( !row._size )
+		return;
+
+	ASSERT ( var < _varsNum );
+	double curCoefficient, ci = 0, realCoefficient;
+	bool tempUpper;
+	unsigned curVar;
+	if ( var == row._lhs )
+		ci = -1;
+	else
+	{
+		for ( unsigned i = 0; i < row._size; ++i )
+			if ( row._row[i]._var == var )
+			{
+				ci = row._row[i]._coefficient;
+				break;
+			}
+	}
+
+	ASSERT( !FloatUtils::isZero( ci ) );
+	std::vector<double> rowCoefficients = std::vector<double>( _rowsNum, 0 );
+	std::vector<double> sum = std::vector<double>( _rowsNum, 0 );
+	std::vector<double> tempBound;
+
+	for ( unsigned i = 0; i < row._size; ++i )
+	{
+		curVar = row._row[i]._var;
+		curCoefficient = row._row[i]._coefficient;
+		if ( FloatUtils::isZero( curCoefficient ) || curVar == var ) // If coefficient is zero then nothing to add to the sum, also skip var
+			continue;
+
+		realCoefficient = curCoefficient / -ci;
+		if ( FloatUtils::isZero( realCoefficient ) )
+			continue;
+
+		tempUpper = ( isUpper && realCoefficient > 0 ) || ( !isUpper &&  realCoefficient < 0 ); // If coefficient of lhs and var are different, use same bound
+		tempBound = tempUpper ? _upperBoundsExplanations[curVar] : _lowerBoundsExplanations[curVar];
+		addVecTimesScalar( sum, tempBound, realCoefficient );
+	}
+
+	// Include lhs as well, if needed
+	if ( var != row._lhs )
+	{
+		realCoefficient = 1 / ci;
+		if ( !FloatUtils::isZero( realCoefficient ) )
+		{
+			tempUpper = ( isUpper && realCoefficient > 0 ) || ( !isUpper &&  realCoefficient < 0 ); // If coefficient of lhs and var are different, use same bound
+			tempBound = tempUpper ? _upperBoundsExplanations[row._lhs] : _lowerBoundsExplanations[row._lhs];
+			addVecTimesScalar( sum, tempBound, realCoefficient );
+		}
+	}
+
+	extractRowCoefficients( row, rowCoefficients, ci ); // Update according to row coefficients
+	addVecTimesScalar( sum, rowCoefficients, 1 );
+	injectExplanation( sum, var, isUpper );
+
```

Comment:
Doesn't inject override sum, that we populated just now?

Reply:
Inject updates sum to be an explanation of var, without changing it.